// @mlightcad/shx-parser/dist/index.es.js
class f {
  /**
   * Converts an unsigned byte to a signed byte as used in SHX format.
   * Values > 127 are converted to their signed equivalent (-128 to -1).
   * @param value - The unsigned byte value to convert
   * @returns The signed byte value
   */
  static byteToSByte(t) {
    return (t & 127) - (t & 128 ? 128 : 0);
  }
  /**
   * Creates a new ShxFileReader instance.
   * @param arraybuffer - The ArrayBuffer to read from
   */
  constructor(t) {
    this.position = 0, this.data = new DataView(t);
  }
  /**
   * Reads a specified number of bytes from the current position.
   * @param length - Number of bytes to read (optional)
   * @returns A Uint8Array containing the read bytes
   * @throws Error if reading beyond buffer bounds
   */
  readBytes(t = 1) {
    this.data.byteLength < this.position + t && this.throwOutOfRangeError(this.position + t);
    const e = new Uint8Array(this.data.buffer, this.position, t);
    return this.position += t, e;
  }
  /**
   * Skips a specified number of bytes from the current position.
   * @param length - Number of bytes to skip
   * @throws Error if skipping beyond buffer bounds
   */
  skip(t) {
    this.data.byteLength < this.position + t && this.throwOutOfRangeError(this.position + t), this.position += t;
  }
  /**
   * Reads an unsigned 8-bit integer.
   * @returns The read uint8 value
   * @throws Error if reading beyond buffer bounds
   */
  readUint8() {
    this.data.byteLength < this.position + 1 && this.throwOutOfRangeError(this.position + 1);
    const t = this.data.getUint8(this.position);
    return this.position += 1, t;
  }
  /**
   * Reads a signed 8-bit integer.
   * @returns The read int8 value
   * @throws Error if reading beyond buffer bounds
   */
  readInt8() {
    this.data.byteLength < this.position + 1 && this.throwOutOfRangeError(this.position + 1);
    const t = this.data.getInt8(this.position);
    return this.position += 1, t;
  }
  /**
   * Reads an unsigned 16-bit integer.
   * @param littleEndian If false, a big-endian value should be read.
   * @returns The read uint16 value
   * @throws Error if reading beyond buffer bounds
   */
  readUint16(t = !0) {
    this.data.byteLength < this.position + 2 && this.throwOutOfRangeError(this.position + 2);
    const e = this.data.getUint16(this.position, t);
    return this.position += 2, e;
  }
  /**
   * Reads a signed 16-bit integer.
   * @returns The read int16 value
   * @throws Error if reading beyond buffer bounds
   */
  readInt16() {
    this.data.byteLength < this.position + 2 && this.throwOutOfRangeError(this.position + 2);
    const t = this.data.getInt16(this.position, !0);
    return this.position += 2, t;
  }
  /**
   * Reads an unsigned 32-bit integer.
   * @returns The read uint32 value
   * @throws Error if reading beyond buffer bounds
   */
  readUint32() {
    this.data.byteLength < this.position + 4 && this.throwOutOfRangeError(this.position + 4);
    const t = this.data.getUint32(this.position, !0);
    return this.position += 4, t;
  }
  /**
   * Reads a signed 32-bit integer.
   * @returns The read int32 value
   * @throws Error if reading beyond buffer bounds
   */
  readInt32() {
    this.data.byteLength < this.position + 4 && this.throwOutOfRangeError(this.position + 4);
    const t = this.data.getInt32(this.position, !0);
    return this.position += 4, t;
  }
  /**
   * Reads a 32-bit floating point number.
   * @returns The read float32 value
   * @throws Error if reading beyond buffer bounds
   */
  readFloat32() {
    this.data.byteLength < this.position + 4 && this.throwOutOfRangeError(this.position + 4);
    const t = this.data.getFloat32(this.position, !0);
    return this.position += 4, t;
  }
  /**
   * Reads a 64-bit floating point number.
   * @returns The read float64 value
   * @throws Error if reading beyond buffer bounds
   */
  readFloat64() {
    this.data.byteLength < this.position + 8 && this.throwOutOfRangeError(this.position + 8);
    const t = this.data.getFloat64(this.position, !0);
    return this.position += 8, t;
  }
  /**
   * Sets the current read position in the buffer.
   * @param position - The new position to set
   */
  setPosition(t) {
    this.data.byteLength < t && this.throwOutOfRangeError(t), this.position = t;
  }
  /**
   * Checks if the current position is at the end of the buffer.
   * @returns True if at the end of the buffer, false otherwise
   */
  isEnd() {
    return this.position === this.data.byteLength - 1;
  }
  /**
   * Gets the current position in the buffer.
   * @returns The current position
   */
  get currentPosition() {
    return this.position;
  }
  /**
   * Gets the total length of the buffer.
   * @returns The buffer length in bytes
   */
  get length() {
    return this.data.byteLength;
  }
  /**
   * Throws an error when attempting to read beyond buffer bounds.
   * @param position - The position that caused the error
   * @throws Error with details about the out of range access
   */
  throwOutOfRangeError(t) {
    throw new Error(
      `Position ${t} is out of range for the data length ${this.data.byteLength}!`
    );
  }
}
var g = /* @__PURE__ */ ((b) => (b.SHAPES = "shapes", b.BIGFONT = "bigfont", b.UNIFONT = "unifont", b))(g || {});
class I {
  parse(t) {
    const i = this.parseHeader(t).split(" "), n = i[1].toLocaleLowerCase();
    if (!Object.values(g).includes(n))
      throw new Error(`Invalid font type: ${n}`);
    return {
      fileHeader: i[0],
      fontType: n,
      fileVersion: i[2]
    };
  }
  parseHeader(t) {
    let e = "", n = 0;
    for (; t.currentPosition < t.length - 2 && n < 1024; ) {
      const s = t.readUint8();
      if (s === 13) {
        const o = t.currentPosition, r = t.readUint8(), a = t.readUint8();
        if (r === 10 && a === 26)
          break;
        t.setPosition(o), e += String.fromCharCode(s);
      } else
        e += String.fromCharCode(s);
      n++;
    }
    return e.trim();
  }
}
const w = 10, B = [13, 10, 0];
class T {
  parse(t) {
    try {
      t.readBytes(4);
      const e = t.readInt16();
      if (e <= 0)
        throw new Error("Invalid shape count in font file");
      const i = [];
      for (let o = 0; o < e; o++) {
        const r = t.readUint16(), a = t.readUint16();
        a > 0 && i.push({ code: r, length: a });
      }
      const n = {};
      for (const o of i)
        try {
          const r = t.readBytes(o.length);
          r.length === o.length && (n[o.code] = r);
        } catch {
          console.warn(`Failed to read shape data for code ${o.code}`);
        }
      const s = {
        data: n,
        info: "",
        baseUp: 8,
        baseDown: 2,
        height: w,
        width: w,
        orientation: "horizontal",
        isExtended: !1
      };
      if (0 in n) {
        const o = n[0];
        try {
          const r = new TextDecoder().decode(o);
          let a = o.findIndex((h) => B.includes(h));
          a >= 0 && (s.info = r.substring(0, a), a + 3 < o.length && (s.baseUp = o[a + 1], s.baseDown = o[a + 2], s.height = s.baseDown + s.baseUp, s.width = s.height, s.orientation = o[a + 3] === 0 ? "horizontal" : "vertical"));
        } catch {
          console.warn("Failed to parse font info block");
        }
      }
      return s;
    } catch (e) {
      return console.error("Error parsing shape font:", e), {
        data: {},
        info: "Failed to parse font file",
        baseUp: 8,
        baseDown: 2,
        height: w,
        width: w,
        orientation: "horizontal",
        isExtended: !1
      };
    }
  }
}
class U {
  parse(t) {
    try {
      t.readInt16();
      const e = t.readInt16(), i = t.readInt16();
      if (e <= 0)
        throw new Error("Invalid character count in font file");
      t.skip(i * 4);
      const n = [];
      for (let r = 0; r < e; r++) {
        const a = t.readUint16(), h = t.readUint16(), c = t.readUint32();
        (a !== 0 || h !== 0 || c !== 0) && n.push({ code: a, length: h, offset: c });
      }
      const s = {};
      for (const r of n)
        try {
          t.setPosition(r.offset);
          const a = t.readBytes(r.length);
          a.length === r.length && (s[r.code] = a);
        } catch {
          console.warn(`Failed to read bigfont data for code ${r.code}`);
        }
      const o = {
        data: s,
        info: "",
        baseUp: 8,
        baseDown: 2,
        height: w,
        width: w,
        orientation: "horizontal",
        isExtended: !1
      };
      if (0 in s) {
        const r = s[0];
        try {
          const a = this.utf8ArrayToStr(r);
          let h = a.pos;
          h >= 0 && (o.info = a.text, h++, h + 3 < r.length && (r.length - h > 4 ? (o.height = r[h++], h++, o.orientation = r[h++] === 0 ? "horizontal" : "vertical", o.width = r[h++], o.baseUp = o.height, o.baseDown = 0, o.isExtended = !0) : (o.baseUp = r[h++], o.baseDown = r[h++], o.height = o.baseDown + o.baseUp, o.width = o.height, o.orientation = r[h] === 0 ? "horizontal" : "vertical")));
        } catch {
          console.warn("Failed to parse bigfont info block");
        }
      }
      return o;
    } catch (e) {
      return console.error("Error parsing big font:", e), {
        data: {},
        info: "Failed to parse font file",
        baseUp: 8,
        baseDown: 2,
        height: w,
        width: w,
        orientation: "horizontal",
        isExtended: !1
      };
    }
  }
  utf8ArrayToStr(t) {
    let e = "", i = 0;
    for (; i < t.length; ) {
      const n = t[i];
      switch (n >> 4) {
        case 0:
        case 1:
        case 2:
        case 3:
        case 4:
        case 5:
        case 6:
        case 7:
          e += String.fromCharCode(n);
          break;
        case 12:
        case 13: {
          const s = t[i++];
          e += String.fromCharCode((n & 31) << 6 | s & 63);
          break;
        }
        case 14: {
          const s = t[i++], o = t[i++];
          e += String.fromCharCode(
            (n & 15) << 12 | (s & 63) << 6 | (o & 63) << 0
          );
          break;
        }
      }
      if (e.charCodeAt(e.length - 1) === 0) break;
      i++;
    }
    return { text: e, pos: i };
  }
}
class F {
  parse(t) {
    try {
      const e = t.readInt32();
      if (e <= 0)
        throw new Error("Invalid character count in font file");
      const i = t.readInt16(), n = t.readBytes(i), s = {
        data: {},
        info: "",
        baseUp: 8,
        baseDown: 2,
        height: w,
        width: w,
        orientation: "horizontal",
        isExtended: !1
      };
      try {
        const r = new TextDecoder().decode(n);
        let a = r.indexOf("\0");
        a >= 0 && (s.info = r.substring(0, a), a + 3 < n.length && (s.baseUp = n[a + 1], s.baseDown = n[a + 2], s.height = s.baseUp + s.baseDown, s.width = s.height, s.orientation = n[a + 3] === 0 ? "horizontal" : "vertical"));
      } catch {
        console.warn("Failed to parse unifont info block");
      }
      const o = {};
      for (let r = 0; r < e - 1; r++)
        try {
          const a = t.readUint16(), h = t.readUint16();
          if (h > 0) {
            const c = t.readBytes(h);
            if (c.length === h) {
              const u = c.indexOf(0);
              let d = 0;
              u >= 0 && u < c.length && (d = u + 1), d < c.length && (o[a] = c.subarray(d));
            }
          }
        } catch {
          console.warn("Failed to read unifont character data");
          break;
        }
      return s.data = o, s;
    } catch (e) {
      return console.error("Error parsing unifont:", e), {
        data: {},
        info: "Failed to parse font file",
        baseUp: 8,
        baseDown: 2,
        height: w,
        width: w,
        orientation: "horizontal",
        isExtended: !1
      };
    }
  }
}
class v {
  static createParser(t) {
    switch (t) {
      case g.SHAPES:
        return new T();
      case g.BIGFONT:
        return new U();
      case g.UNIFONT:
        return new F();
      default:
        throw new Error(`Unsupported font type: ${t}`);
    }
  }
}
class l {
  /**
   * Creates a new Point instance.
   * @param x - The x-coordinate (defaults to 0)
   * @param y - The y-coordinate (defaults to 0)
   */
  constructor(t = 0, e = 0) {
    this.x = t, this.y = e;
  }
  /**
   * Sets the coordinates of the point.
   * @param x - The new x-coordinate
   * @param y - The new y-coordinate
   * @returns The point instance for method chaining
   */
  set(t, e) {
    return this.x = t, this.y = e, this;
  }
  /**
   * Calculates the length (magnitude) of the vector from origin to this point.
   * @returns The length of the vector
   */
  length() {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }
  /**
   * Normalizes the point vector to have a length of 1.
   * @returns The point instance for method chaining
   */
  normalize() {
    const t = this.length();
    return t !== 0 && (this.x /= t, this.y /= t), this;
  }
  /**
   * Creates a new Point instance with the same coordinates.
   * @returns A new Point instance with the same x and y values
   */
  clone() {
    return new l(this.x, this.y);
  }
  /**
   * Adds another point's coordinates to this point.
   * @param point - The point to add
   * @returns The point instance for method chaining
   */
  add(t) {
    return this.x += t.x, this.y += t.y, this;
  }
  /**
   * Subtracts another point's coordinates from this point.
   * @param point - The point to subtract
   * @returns The point instance for method chaining
   */
  subtract(t) {
    return this.x -= t.x, this.y -= t.y, this;
  }
  /**
   * Multiplies both coordinates by a scalar value.
   * @param scalar - The scalar value to multiply by
   * @returns The point instance for method chaining
   */
  multiply(t) {
    return this.x *= t, this.y *= t, this;
  }
  /**
   * Divides both coordinates by a scalar value.
   * @param scalar - The scalar value to divide by
   * @returns The point instance for method chaining
   */
  divide(t) {
    return t !== 0 && (this.x /= t, this.y /= t), this;
  }
  /**
   * Multiplies x and y coordinates by different scalar values.
   * @param xScalar - The scalar value to multiply x-coordinate by
   * @param yScalar - The scalar value to multiply y-coordinate by
   * @returns The point instance for method chaining
   */
  multiplyScalars(t, e) {
    return this.x *= t, this.y *= e, this;
  }
  /**
   * Divides x and y coordinates by different scalar values.
   * @param xScalar - The scalar value to divide x-coordinate by
   * @param yScalar - The scalar value to divide y-coordinate by
   * @returns The point instance for method chaining
   */
  divideScalars(t, e) {
    return t !== 0 && (this.x /= t), e !== 0 && (this.y /= e), this;
  }
  /**
   * Calculates the Euclidean distance to another point.
   * @param point - The point to calculate distance to
   * @returns The distance between the two points
   */
  distanceTo(t) {
    const e = this.x - t.x, i = this.y - t.y;
    return Math.sqrt(e * e + i * i);
  }
}
const E = Math.PI / 4;
class C {
  /**
   * Creates a bulge-defined arc
   * @param start Start point
   * @param end End point
   * @param bulge Bulge factor (-1 to 1, where 1 is a semicircle)
   */
  static fromBulge(t, e, i) {
    const n = Math.max(-1, Math.min(1, i));
    return new C({
      start: t,
      end: e,
      bulge: n
    });
  }
  /**
   * Creates an octant-defined arc
   * @param center Center point of the arc
   * @param radius Radius of the arc
   * @param startOctant Starting octant (0-7)
   * @param octantCount Number of octants to span (0-8, where 0 means 8 octants)
   * @param isClockwise Whether the arc goes clockwise
   */
  static fromOctant(t, e, i, n, s) {
    return new C({
      center: t,
      radius: e,
      startOctant: i,
      octantCount: n,
      isClockwise: s
    });
  }
  constructor(t) {
    if (t.start && t.end && t.bulge !== void 0) {
      this.start = t.start.clone(), this.end = t.end.clone(), this.bulge = t.bulge, this.isClockwise = t.bulge < 0;
      const e = this.end.clone().subtract(this.start), i = e.length();
      if (Math.abs(this.bulge) * i / 2 === 0) {
        this.radius = 0, this.center = this.start.clone(), this.startAngle = Math.atan2(e.y, e.x), this.endAngle = this.startAngle;
        return;
      }
      const s = 4 * Math.atan(Math.abs(this.bulge));
      this.radius = i / (2 * Math.sin(s / 2));
      const o = this.start.clone().add(e.clone().divide(2)), r = new l(-e.y, e.x);
      r.normalize(), r.multiply(Math.abs(this.radius * Math.cos(s / 2))), this.center = o.clone(), this.isClockwise ? this.center.subtract(r) : this.center.add(r), this.startAngle = Math.atan2(this.start.y - this.center.y, this.start.x - this.center.x), this.endAngle = Math.atan2(this.end.y - this.center.y, this.end.x - this.center.x), this.isClockwise ? this.endAngle >= this.startAngle && (this.endAngle -= 2 * Math.PI) : this.endAngle <= this.startAngle && (this.endAngle += 2 * Math.PI);
    } else if (t.center && t.radius !== void 0 && t.startOctant !== void 0 && t.octantCount !== void 0 && t.isClockwise !== void 0) {
      this.center = t.center.clone(), this.radius = t.radius, this.isClockwise = t.isClockwise, this.startAngle = t.startOctant * E;
      const e = (t.octantCount === 0 ? 8 : t.octantCount) * E;
      this.endAngle = this.startAngle + (this.isClockwise ? -e : e), this.start = this.center.clone().add(
        new l(
          this.radius * Math.cos(this.startAngle),
          this.radius * Math.sin(this.startAngle)
        )
      ), this.end = this.center.clone().add(
        new l(this.radius * Math.cos(this.endAngle), this.radius * Math.sin(this.endAngle))
      );
    } else
      throw new Error("Invalid arc parameters");
  }
  /**
   * Tessellates the arc into a series of points that approximate the arc.
   * @param circleSpan The angle span between tessellated points (default Math.PI / 18)
   * @returns Array of points representing the tessellated arc
   */
  tessellate(t = Math.PI / 18) {
    if (this.radius === 0)
      return [this.start.clone(), this.end.clone()];
    const e = [this.start.clone()], i = Math.abs(this.endAngle - this.startAngle), n = Math.max(1, Math.floor(i / t));
    for (let s = 1; s < n; s++) {
      const o = s / n, r = this.isClockwise ? this.startAngle - o * i : this.startAngle + o * i;
      e.push(
        this.center.clone().add(new l(this.radius * Math.cos(r), this.radius * Math.sin(r)))
      );
    }
    return e.push(
      this.end ? this.end.clone() : this.center.clone().add(
        new l(
          this.radius * Math.cos(this.endAngle),
          this.radius * Math.sin(this.endAngle)
        )
      )
    ), e;
  }
}
class A {
  constructor(t, e = []) {
    this.lastPoint = t, this.polylines = e;
  }
  /**
   * Get the bounding box of the shape
   * @returns Bounding box of the shape
   */
  get bbox() {
    if (this._bbox)
      return this._bbox;
    let t = 1 / 0, e = -1 / 0, i = 1 / 0, n = -1 / 0;
    return this.polylines.forEach((s) => {
      s.forEach((o) => {
        t = Math.min(t, o.x), e = Math.max(e, o.x), i = Math.min(i, o.y), n = Math.max(n, o.y);
      });
    }), this._bbox = { minX: t, minY: i, maxX: e, maxY: n }, this._bbox;
  }
  /**
   * Offset the shape by a point
   * @param p The point to offset the shape by
   * @param isNewInstance Whether to return a new instance of the shape or modify the current instance
   * @returns The offset shape
   */
  offset(t, e = !0) {
    var i, n;
    return e ? new A(
      (i = this.lastPoint) == null ? void 0 : i.clone().add(t),
      this.polylines.map((s) => s.map((o) => o.clone().add(t)))
    ) : ((n = this.lastPoint) == null || n.add(t), this.polylines.forEach((s) => s.forEach((o) => o.add(t))), this._bbox && (this._bbox.maxX += t.x, this._bbox.minX += t.x, this._bbox.maxY += t.y, this._bbox.minY += t.y), this);
  }
  /**
   * Normalizes a shape so that its bounding box’s bottom-left corner moves to the origin (0,0).
   * It doesn’t change the size or orientation, only repositions the shape.
   * @param isNewInstance Whether to return a new instance of the shape or modify the current instance
   * @returns The offset shape
   */
  normalizeToOrigin(t = !1) {
    const e = this.bbox;
    return this.offset(new l(-e.minX, -e.minY), t);
  }
  /**
   * Converts the shape to an SVG string
   * @param options SVG rendering options
   * @returns SVG string
   */
  toSVG(t = {}) {
    const { strokeWidth: e = "0.5%", strokeColor: i = "black", isAutoFit: n = !1 } = t;
    let s, o;
    if (n) {
      const r = this.bbox, a = 0.2, h = r.maxX - r.minX, c = r.maxY - r.minY, u = h === 0 ? c : h, d = c === 0 ? h : c, p = r.minX - u * a, x = r.maxX + u * a, k = r.minY - d * a, M = r.maxY + d * a;
      o = this.polylines.map((D) => {
        let P = "";
        return D.forEach((m, y) => {
          const S = m.x, O = -m.y;
          P += y === 0 ? `M ${S} ${O} ` : `L ${S} ${O} `;
        }), `<path d="${P}" stroke="${i}" stroke-width="${e}" fill="none"/>`;
      }).join(""), s = `${p} ${-M} ${x - p} ${M - k}`;
    } else
      s = "0 0 20 20", o = this.polylines.map((r) => {
        let a = "";
        return r.forEach((h, c) => {
          const u = h.x + 5, d = -h.y + 15;
          a += c === 0 ? `M ${u} ${d} ` : `L ${u} ${d} `;
        }), `<path d="${a}" stroke="${i}" stroke-width="${e}" fill="none"/>`;
      }).join("");
    return `<svg width="100%" height="100%" viewBox="${s}" preserveAspectRatio="xMidYMid meet">${o}</svg>`;
  }
}
const L = Math.PI / 18;
class $ {
  constructor(t) {
    this.shapeCache = /* @__PURE__ */ new Map(), this.shapeData = /* @__PURE__ */ new Map(), this.fontData = t;
  }
  /**
   * Releases parsed shapes and cached shapes
   */
  release() {
    this.shapeCache.clear(), this.shapeData.clear();
  }
  /**
   * Parses a character's shape with the given font size.
   * @param code - The character code
   * @param size - The font size
   * @returns The parsed shape or undefined if the character is not found
   */
  getCharShape(t, e) {
    const i = e / this.fontData.content.height;
    return this.parseAndScale(t, { factor: i });
  }
  /**
   * Parses a character's shape with scaling options
   * @param code - The character code
   * @param options - Scaling options (factor or height/width)
   * @returns The parsed shape or undefined if the character is not found
   */
  parseAndScale(t, e) {
    if (t === 0)
      return;
    let i;
    if (this.shapeCache.has(t))
      i = this.shapeCache.get(t);
    else {
      const n = this.fontData.content.data;
      if (n[t]) {
        const s = n[t];
        i = this.parseShape(s), this.shapeData.set(t, i), this.shapeCache.set(t, i);
      }
    }
    if (i) {
      if (e.factor !== void 0)
        return this.scaleShapeByFactor(i, e.factor);
      if (e.height !== void 0) {
        const n = e.width ?? e.height;
        return this.scaleShapeByHeightAndWidth(i, e.height, n);
      } else
        return i;
    }
  }
  /**
   * Scales a shape according to the given scale factor
   * @param shape - The shape to scale
   * @param factor - The scale factor
   * @returns The scaled shape
   */
  scaleShapeByFactor(t, e) {
    var i;
    return new A(
      (i = t.lastPoint) == null ? void 0 : i.clone().multiply(e),
      t.polylines.map((n) => n.map((s) => s.clone().multiply(e)))
    );
  }
  /**
   * Scales a shape according to the given height and width
   * @param shape - The shape to scale
   * @param height - The target height
   * @param width - The target width
   * @returns The scaled shape
   */
  scaleShapeByHeightAndWidth(t, e, i) {
    var u;
    const n = t.bbox, s = n.maxY - n.minY, o = n.maxX - n.minX, r = s > 0 ? e / s : 1, a = o > 0 ? i / o : 1, h = (u = t.lastPoint) == null ? void 0 : u.clone();
    h && (h.x *= a, h.y *= r);
    const c = t.polylines.map(
      (d) => d.map((p) => {
        const x = p.clone();
        return x.x *= a, x.y *= r, x;
      })
    );
    return new A(h, c);
  }
  /**
   * Parses the shape of a character.
   * @param data - The data of the character
   * @returns The parsed shape
   */
  parseShape(t) {
    const r = {
      currentPoint: new l(),
      polylines: [],
      currentPolyline: [],
      sp: [],
      isPenDown: !1,
      scale: 1
    };
    for (let a = 0; a < t.length; a++) {
      const h = t[a];
      h <= 15 ? a = this.handleSpecialCommand(h, t, a, r) : this.handleVectorCommand(h, r);
    }
    return new A(r.currentPoint, r.polylines);
  }
  /**
   * Please refer to special codes reference in the following link for more information.
   * https://help.autodesk.com/view/OARX/2023/ENU/?guid=GUID-06832147-16BE-4A66-A6D0-3ADF98DC8228
   * @param command - The command byte
   * @param data - The data of the character
   * @param index - The index of the command byte
   * @param state - The state of the parser
   * @returns The index of the next command byte
   */
  handleSpecialCommand(t, e, i, n) {
    let s = i;
    switch (t) {
      case 0:
        n.currentPolyline = [], n.isPenDown = !1;
        break;
      case 1:
        n.isPenDown = !0, n.currentPolyline.push(n.currentPoint.clone());
        break;
      case 2:
        n.isPenDown = !1, n.currentPolyline.length > 1 && n.polylines.push(n.currentPolyline.slice()), n.currentPolyline = [];
        break;
      case 3:
        s++, n.scale /= e[s];
        break;
      case 4:
        s++, n.scale *= e[s];
        break;
      case 5:
        if (n.sp.length === 4)
          throw new Error("The position stack is only four locations deep");
        n.sp.push(n.currentPoint.clone());
        break;
      case 6:
        n.currentPoint = n.sp.pop() ?? n.currentPoint;
        break;
      case 7:
        s = this.handleSubshapeCommand(e, s, n);
        break;
      case 8:
        s = this.handleXYDisplacement(e, s, n);
        break;
      case 9:
        s = this.handleMultipleXYDisplacements(e, s, n);
        break;
      case 10:
        s = this.handleOctantArc(e, s, n);
        break;
      case 11:
        s = this.handleFractionalArc(e, s, n);
        break;
      case 12:
        s = this.handleBulgeArc(e, s, n);
        break;
      case 13:
        s = this.handleMultipleBulgeArcs(e, s, n);
        break;
      case 14:
        s = this.skipCode(e, ++s);
        break;
    }
    return s;
  }
  handleVectorCommand(t, e) {
    const i = (t & 240) >> 4, n = t & 15, s = this.getVectorForDirection(n);
    e.currentPoint.add(s.multiply(i * e.scale)), e.isPenDown && e.currentPolyline.push(e.currentPoint.clone());
  }
  /**
   * Get the vector for the given direction code. Please refer to the following link for more information.
   * https://help.autodesk.com/view/OARX/2023/ENU/?guid=GUID-0A8E12A1-F4AB-44AD-8A9B-2140E0D5FD23
   * @param dir - The direction code of the vector
   * @returns Returns the vector for the given direction code
   */
  getVectorForDirection(t) {
    const e = new l();
    switch (t) {
      case 0:
        e.x = 1;
        break;
      case 1:
        e.x = 1, e.y = 0.5;
        break;
      case 2:
        e.x = 1, e.y = 1;
        break;
      case 3:
        e.x = 0.5, e.y = 1;
        break;
      case 4:
        e.y = 1;
        break;
      case 5:
        e.x = -0.5, e.y = 1;
        break;
      case 6:
        e.x = -1, e.y = 1;
        break;
      case 7:
        e.x = -1, e.y = 0.5;
        break;
      case 8:
        e.x = -1;
        break;
      case 9:
        e.x = -1, e.y = -0.5;
        break;
      case 10:
        e.x = -1, e.y = -1;
        break;
      case 11:
        e.x = -0.5, e.y = -1;
        break;
      case 12:
        e.y = -1;
        break;
      case 13:
        e.x = 0.5, e.y = -1;
        break;
      case 14:
        e.x = 1, e.y = -1;
        break;
      case 15:
        e.x = 1, e.y = -0.5;
        break;
    }
    return e;
  }
  handleSubshapeCommand(t, e, i) {
    let n = e, s = 0, o, r = i.scale * this.fontData.content.height, a = r;
    const h = i.currentPoint.clone();
    switch (i.currentPolyline.length > 1 && (i.polylines.push(i.currentPolyline.slice()), i.currentPolyline = []), this.fontData.header.fontType) {
      case g.SHAPES:
        n++, s = t[n];
        break;
      case g.BIGFONT:
        n++, s = t[n], s === 0 && (n++, s = t[n++] << 8 | t[n++], h.x = f.byteToSByte(t[n++]) * i.scale, h.y = f.byteToSByte(t[n++]) * i.scale, this.fontData.content.isExtended && (a = t[n++] * i.scale), r = t[n] * i.scale);
        break;
      case g.UNIFONT:
        n++, s = t[n++] << 8 | t[n++];
        break;
    }
    return s !== 0 && (o = this.getScaledSubshapeAtInsertPoint(s, a, r, h), o && i.polylines.push(...o.polylines.slice())), i.currentPolyline = [], n;
  }
  handleXYDisplacement(t, e, i) {
    let n = e;
    const s = new l();
    return s.x = f.byteToSByte(t[++n]), s.y = f.byteToSByte(t[++n]), i.currentPoint.add(s.multiply(i.scale)), i.isPenDown && i.currentPolyline.push(i.currentPoint.clone()), n;
  }
  handleMultipleXYDisplacements(t, e, i) {
    let n = e;
    for (; ; ) {
      const s = new l();
      if (s.x = f.byteToSByte(t[++n]), s.y = f.byteToSByte(t[++n]), s.x === 0 && s.y === 0)
        break;
      i.currentPoint.add(s.multiply(i.scale)), i.isPenDown && i.currentPolyline.push(i.currentPoint.clone());
    }
    return n;
  }
  handleOctantArc(t, e, i) {
    var p;
    let n = e;
    const s = t[++n] * i.scale, o = f.byteToSByte(t[++n]), r = (o & 112) >> 4;
    let a = o & 7;
    const h = o < 0, c = Math.PI / 4 * r, u = i.currentPoint.clone().subtract(new l(Math.cos(c) * s, Math.sin(c) * s)), d = C.fromOctant(u, s, r, a, h);
    if (i.isPenDown) {
      const x = d.tessellate();
      i.currentPolyline.pop(), i.currentPolyline.push(...x.slice());
    }
    return i.currentPoint = (p = d.tessellate().pop()) == null ? void 0 : p.clone(), n;
  }
  handleFractionalArc(t, e, i) {
    let n = e;
    const s = t[++n], o = t[++n], r = t[++n], a = t[++n], h = (r * 255 + a) * i.scale, c = f.byteToSByte(t[++n]), u = (c & 112) >> 4;
    let d = c & 7;
    d === 0 && (d = 8), o !== 0 && d--;
    const p = Math.PI / 4;
    let x = p * d, k = L, M = 1;
    c < 0 && (k = -k, x = -x, M = -1);
    let D = p * u, P = D + x;
    D += p * s / 256 * M, P += p * o / 256 * M;
    const m = i.currentPoint.clone().subtract(new l(h * Math.cos(D), h * Math.sin(D)));
    if (i.currentPoint = m.clone().add(new l(h * Math.cos(P), h * Math.sin(P))), i.isPenDown) {
      let y = D;
      const S = [];
      if (S.push(
        m.clone().add(new l(h * Math.cos(y), h * Math.sin(y)))
      ), k > 0)
        for (; y + k < P; )
          y += k, S.push(
            m.clone().add(new l(h * Math.cos(y), h * Math.sin(y)))
          );
      else
        for (; y + k > P; )
          y += k, S.push(
            m.clone().add(new l(h * Math.cos(y), h * Math.sin(y)))
          );
      S.push(m.clone().add(new l(h * Math.cos(P), h * Math.sin(P)))), i.currentPolyline.push(...S);
    }
    return n;
  }
  handleBulgeArc(t, e, i) {
    let n = e;
    const s = new l();
    s.x = f.byteToSByte(t[++n]), s.y = f.byteToSByte(t[++n]);
    const o = f.byteToSByte(t[++n]);
    return i.currentPoint = this.handleArcSegment(
      i.currentPoint,
      s,
      o,
      i.scale,
      i.isPenDown,
      i.currentPolyline
    ), n;
  }
  handleMultipleBulgeArcs(t, e, i) {
    let n = e;
    for (; ; ) {
      const s = new l();
      if (s.x = f.byteToSByte(t[++n]), s.y = f.byteToSByte(t[++n]), s.x === 0 && s.y === 0)
        break;
      const o = f.byteToSByte(t[++n]);
      i.currentPoint = this.handleArcSegment(
        i.currentPoint,
        s,
        o,
        i.scale,
        i.isPenDown,
        i.currentPolyline
      );
    }
    return n;
  }
  skipCode(t, e) {
    switch (t[e]) {
      case 0:
        break;
      case 1:
        break;
      case 2:
        break;
      case 3:
      case 4:
        e++;
        break;
      case 5:
        break;
      case 6:
        break;
      case 7:
        switch (this.fontData.header.fontType) {
          case g.SHAPES:
            e++;
            break;
          case g.BIGFONT:
            e++, t[e] === 0 && (e += this.fontData.content.isExtended ? 6 : 5);
            break;
          case g.UNIFONT:
            e += 2;
            break;
        }
        break;
      case 8:
        e += 2;
        break;
      case 9:
        for (; ; ) {
          const s = t[++e], o = t[++e];
          if (s === 0 && o === 0)
            break;
        }
        break;
      case 10:
        e += 2;
        break;
      case 11:
        e += 5;
        break;
      case 12:
        e += 3;
        break;
      case 13:
        for (; ; ) {
          const s = t[++e], o = t[++e];
          if (s === 0 && o === 0)
            break;
          e++;
        }
        break;
    }
    return e;
  }
  getScaledSubshapeAtInsertPoint(t, e, i, n) {
    let s = this.shapeCache.get(t);
    if (!s) {
      const a = this.fontData.content.data[t];
      if (!a)
        return;
      s = this.parseShape(a), this.shapeData.set(t, s), this.shapeCache.set(t, s);
    }
    const o = s.normalizeToOrigin(!0);
    return this.scaleShapeByHeightAndWidth(o, i, e).offset(n, !1);
  }
  /**
   * Handles drawing an arc segment with the given vector and bulge
   * @param currentPoint The starting point of the arc
   * @param vec The displacement vector
   * @param bulge The bulge value (will be normalized by 127.0)
   * @param scale The current scale factor
   * @param isPenDown Whether the pen is currently down (drawing)
   * @param currentPolyline The current polyline being built
   * @returns The new current point after the arc
   */
  handleArcSegment(t, e, i, n, s, o) {
    e.x *= n, e.y *= n, i < -127 && (i = -127);
    const r = t.clone();
    if (s)
      if (i === 0)
        o.push(r.clone().add(e));
      else {
        const a = r.clone().add(e), c = C.fromBulge(r, a, i / 127).tessellate();
        o.push(...c.slice(1));
      }
    return r.add(e), r;
  }
}
class R {
  /**
   * Creates a new ShxFont instance.
   * @param data - Either raw binary data of the SHX font file (ArrayBuffer) or pre-parsed font data (ShxFontData)
   * @throws {Error} If the font data is invalid or cannot be parsed
   */
  constructor(t) {
    if (t instanceof ArrayBuffer) {
      const e = new f(t), n = new I().parse(e), o = v.createParser(n.fontType).parse(e);
      this.fontData = {
        header: n,
        content: o
      };
    } else
      this.fontData = t;
    this.shapeParser = new $(this.fontData);
  }
  /**
   * Return true if this font contains glyph of the specified character. Otherwise, return false.
   * @param char - The character to check
   * @returns True if this font contains glyph of the specified character. Otherwise, return false.
   */
  hasChar(t) {
    return this.fontData.content.data[t] !== void 0;
  }
  /**
   * Gets the shape data for a specific character at a given font size.
   * @param code - The character code to get the shape for
   * @param size - The desired font size
   * @returns The shape data for the character, or undefined if the character is not found in the font
   */
  getCharShape(t, e) {
    let i = this.shapeParser.getCharShape(t, e);
    return i && this.fontData.header.fontType === g.BIGFONT && (i = i.normalizeToOrigin()), i;
  }
  /**
   * Releases resources used by the font.
   * This should be called when the font is no longer needed to free up memory.
   */
  release() {
    this.shapeParser.release();
  }
}
export {
  l as Point,
  R as ShxFont,
  g as ShxFontType,
  A as ShxShape,
  $ as ShxShapeParser
};
